aten/src/ATen/native/cpu/ScatterGatherKernel.cpp

```diff
@@ -275,7 +297,158 @@ void scatter_add_cpu_kernel(Tensor& self, int64_t dim, const Tensor& index, cons
+  auto index_dim_stride = ensure_nonempty_stride(index, dim);
+  auto src_dim_stride = ensure_nonempty_stride(src, dim);
+
+  AT_DISPATCH_ALL_TYPES_AND2(
+    ScalarType::Bool, ScalarType::Half, iter.dtype(),
+    method_name, [&] {
+      using reduce_func_t = std::function<void(scalar_t*, const scalar_t*)>;
+      std::map<const std::string, reduce_func_t> reduce_funcs = {
+          {"sum", reduce_sum},
+          {"subtract", reduce_subtract},
+          {"multiply", reduce_multiply},
+          {"divide", reduce_divide}
+      };
+      auto loop = [&](char** data, const int64_t* strides, int64_t n) {
+        auto* self_data_bytes = data[0];
+        const auto* index_data_bytes = data[2];
+        const auto* src_data_bytes = data[1];
+
+        for (int64_t i = 0; i < n; ++i) {
+          f(
+            (scalar_t*)self_data_bytes, self_dim_stride,
+            (int64_t*)index_data_bytes, index_dim_stride,
+            (scalar_t*)src_data_bytes, src_dim_stride,
+            reduce_funcs, reduce
+          );
+
+          self_data_bytes += strides[0];
+          index_data_bytes += strides[2];
+          src_data_bytes += strides[1];
+        }
+      };
+      if (serial_exec) {
+        iter.serial_for_each(loop, {0, iter.numel()});
+      } else {
+        iter.for_each(loop);
+      }
+    }
+  );
+}
+
+
+void scatter_reduce_cpu_kernel(Tensor& self, int64_t dim, const Tensor& index, const Tensor& src,
+                               std::string& reduce) {
+  if (index.numel() == 0) {
+    return;
+  }
+
+  dim = maybe_wrap_dim(dim, self.dim());
+  
+  scatter_shape_check(self, dim, index, src);
+  int64_t index_dim_size = ensure_nonempty_size(index, dim);
+  int64_t self_dim_size = ensure_nonempty_size(self, dim);
+  std::string method_name = "scatter_" + reduce + "_";
+
+  cpu_scatter_gather_reduce_kernel(
+    self, dim, index, src,
+    method_name, [&] (
+      auto* self_data, auto self_dim_stride,
+      const auto* index_data, auto index_dim_stride,
+      const auto* src_data, auto src_dim_stride,
+      auto& reduce_funcs, const std::string reduce
+    ) {                   
+      for (int64_t i = 0; i < index_dim_size; ++i) {
+        int64_t idx_dim = index_data[i * index_dim_stride];
+        // we are not putting idx_dim in the error message because it disables
+        // loop optimizations in clang-7
+        TORCH_CHECK(idx_dim >= 0 && idx_dim < self_dim_size,
+                    "index ", index_data[i * index_dim_stride],
+                    " is out of bounds for dimension ", dim,
+                    " with size ", self_dim_size);
+
+        std::cout << "inside src: " << src << std::endl;
+        std::cout << "s: " << self_data[idx_dim * self_dim_stride]
+                  << " sr: " << src_data[i * src_dim_stride] << std::endl;
+        reduce_fn(&self_data[idx_dim * self_dim_stride],
+                  &src_data[i * src_dim_stride],
+                  reduce_funcs[reduce]);
```

Comment:
Confusing. I think that passing an operator inside the callback is better? The mapping strings->functor is an internal detail implementation, no need to expose that to a user. Just in my humble opinion :)